portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java

```diff
@@ -341,24 +352,139 @@ public String getPrimaryDomainName() throws UserPortalUIException {
+                        metaClaim, claimValue, offset, length, domainName);
+            } catch (IdentityStoreException e) {
+                String error = "Error while retrieving users for " + claimURI + "= " + claimValue;
+                LOGGER.error(error, e);
+                throw new UserPortalUIException(error);
+            } catch (IllegalStateException e) {
+                String error = e.getMessage();
+                LOGGER.error(error, e);
+                throw new UserPortalUIException(error);
+            }
+        }
+
+        return generateUserListBean(users, requestedClaims);
+
+    }
+
+    @Override
+    public List<UserListBean> getUserList(int offset, int length, String domainName,
+                                          List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<User> users;
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+        try {
+            users = getRealmService().getIdentityStore().listUsers(offset, length, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while retrieving users";
+            LOGGER.error(error, e);
+            throw new UserPortalUIException(error);
+        }
+
+        return generateUserListBean(users, requestedClaims);
+    }
 
     private RealmService getRealmService() {
         if (this.realmService == null) {
             throw new IllegalStateException("Realm Service is null.");
         }
         return this.realmService;
     }
+
+    private List<UserListBean> generateUserListBean(List<User> users, List<ClaimConfigEntry> requestedClaims)
+            throws UserPortalUIException {
+        List<UserListBean> userList = new ArrayList<>();
+
+        Map<String, String> claimDisplayNames = requestedClaims.stream()
+                        .collect(Collectors.toMap(ClaimConfigEntry::getClaimURI, ClaimConfigEntry::getDisplayName));
+
+        List<MetaClaim> metaClaims = requestedClaims.stream()
+                .map(claimConfigEntry -> new MetaClaim("", claimConfigEntry.getClaimURI()))
+                .collect(Collectors.toList());
+
+        List<MetaClaim> groupMetaClaims = new ArrayList<>();
+        MetaClaim groupMetaClaim = new MetaClaim();
+        groupMetaClaim.setClaimUri(GROUPNAME_CLAIM);
+        groupMetaClaims.add(groupMetaClaim);
+
+        for (User user : users) {
+            List<Group> groups;
+            List<Claim> userClaims;
+            List<Claim> groupId;
+            List<String> groupNames = new ArrayList<>();
+            try {
+                groups = user.getGroups();
+                for (Group group : groups) {
+                    //TODO : define groupname as a metadata in groups so this call can be skipped
+                    groupId = group.getClaims(groupMetaClaims);
+                    if (!groupId.isEmpty()) {
+                        groupNames.add(groupId.get(0).getValue());
```

Comment:
Only metaclaim requested is groupname therefore this list contain one claim object. 0th element gives groupname claim